### PR TITLE
tests/vm_test.go: Use libvmi and refactor

### PR DESCRIPTION
**Pull request by 0xFelix**
**What this PR does / why we need it**:

Use libvmi in tests/vm_test.go and do the following refactorings:

- Inline newVirtualMachine and NewRandomVirtualMachineWithRunStrategy
- Rename import of virtctl to allow variables named 'vm'
- Remove duplication of createVirtualMachine()
- Deduplicate VM creation and JSON serialization
- Refactor startVM and stopVM
- Use Eventually where possible
- Use appropriate matchers in virtctl tests
- Consistently name variables across file
- Use const and type inferring where possible
- Fix typos and whitespace
- Replace deprecated call to ioutil.TempFile

**Release note**:

```release-note
None
```

### CODE REVIEW

File: tests/vm_test.go [hunk 1]
```diff
@@ -656,19 +586,10 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
-				if errors.IsNotFound(err) {
-					return false
-				}
-				Expect(err).ToNot(HaveOccurred())
-				return vmi.Status.Phase == v1.Running
-			}, 240*time.Second, 1*time.Second).Should(BeTrue())
+			vmi := libvmi.NewCirros()
+			vmi.Namespace = util.NamespaceTestDefault
+			newVM := createVM(virtClient, vmi, true)
+			vmi = waitForVMIRunning(virtClient, vmi)
```
Comment (rmohr) [inline tests/vm_test.go:592]: Since you touch this already:

You can use `Eventually(ThisVMI(vmi), 120).To(BeRunning())`. That integrates more nicely with the test framework than all those helper functions (like `EventuallyWithContext` can be used) 

We don't have one for `BeReady()` yet, but you could add one.

The benefit is also that it also allows combining conditions ad-hoc when needed, e.g.:


```
Eventually(ThisVMIWith(vm.Namespace, vm.Name), 160).Should(And(BeRunning(), BeOwned()))
```

File: tests/vm_test.go [hunk 2]
```diff
@@ -724,9 +724,7 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 				Eventually(ThisVMIWith(vm.Namespace, vm.Name), 120).Should(BeRunning())
 
 				By("Ensuring a second invocation should fail")
```
Comment (dankenigsberg) [inline tests/vm_test.go:726]: what is the motivation for this change? having shorter code? would you state that in the commit message?

Comment (0xFelix) [reply tests/vm_test.go:726]: The motivation is to use a more appropriate matcher for the error. Instead of using ` Expect(err).To(HaveOccurred())` and `Expect(err.Error()).To(Equal())` you can combine it into `Expect(err).To(MatchError()`. But it shortens the code too.

File: tests/vm_test.go [hunk 3]
```diff
@@ -38,48 +38,52 @@ import (
...
 var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-compute]VirtualMachine", func() {
-
 	var err error
 	var virtClient kubecli.KubevirtClient
 
 	runStrategyAlways := v1.RunStrategyAlways
 	runStrategyHalted := v1.RunStrategyHalted
+	runStrategyManual := v1.RunStrategyManual
+
+	createVM := func(template *v1.VirtualMachineInstance, running bool) *v1.VirtualMachine {
```
Comment (dankenigsberg) [inline tests/vm_test.go:71]: Can this be placed as a top-level `func` tucked in the bottom of this file? Named lambdas are useful when they change or capture a value from withing their closure, but otherwise, the bring about unrequired indentation and complexity.

File: tests/framework/matcher/created.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"fmt"
+
+	"github.com/onsi/gomega/types"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"kubevirt.io/kubevirt/tests/framework/matcher/helper"
+)
+
+func BeCreated() types.GomegaMatcher {
```
Comment (rmohr) [inline tests/framework/matcher/created.go:12]: The matcher as such is great but `BeCreated` may be too close to `Exists` matchers, which would assert if the object passed to `Expect` exists.

I did not check how often this matcher is needed. If often, let's just come up with a nicer name.

If less often, it may make sense to create a general-purpose `unstructured` matcher which could be called like this:


```
Expect(ThisVM(vm)).To(MatchUnstructured("status", "created", true))
```

or something along that line.

File: tests/vm_test.go [hunk 5]
```diff
@@ -769,10 +591,12 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
 		DescribeTable("should report an error status when VM scheduling error occurs", func(unschedulableFunc func(vmi *v1.VirtualMachineInstance)) {
-			vmi := tests.NewRandomVMIWithEphemeralDisk("no-such-image")
+			vmi := libvmi.New(
+				libvmi.WithContainerImage("no-such-image"),
+				libvmi.WithResourceMemory("128Mi"),
+			)
 			unschedulableFunc(vmi)
-
-			vm := createVirtualMachine(true, vmi)
+			vm := createVM(vmi, true)
```
Comment (dankenigsberg) [inline tests/vm_test.go:599]: boolean arguments are problematic to the reader. what's so "true" here, they may ask. A call to `createRunningVM` would be clearer.

File: tests/vm_test.go [hunk 6]
```diff
@@ -1778,20 +1795,3 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
-func getExpectedPodName(vm *v1.VirtualMachine) string {
-	maxNameLength := 63
-	podNamePrefix := "virt-launcher-"
-	podGeneratedSuffixLen := 5
-	charCountFromName := maxNameLength - len(podNamePrefix) - podGeneratedSuffixLen
-	expectedPodName := fmt.Sprintf(fmt.Sprintf("virt-launcher-%%.%ds", charCountFromName), vm.GetName())
-	return expectedPodName
-}
-
-func waitForResourceDeletion(k8sClient string, resourceType string, resourceName string) {
```
Comment (dankenigsberg) [inline tests/vm_test.go:2288]: I would vote against this commit. passing `k8sClient` as an arg is cumbersome, but it's generally better to make the dependency on an almost-global variable explicit. It leaves less room for surprises if somewhere else in the code changes it.

**Comment by 0xFelix** [comment]
@dankenigsberg @rmohr I addressed your comments, WDYT?

File: tests/framework/matcher/restarted.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+		oldUID: oldUID,
+	}
+}
+
+type restartedMatcher struct {
+	oldUID k8stypes.UID
+}
+
+func (r restartedMatcher) Match(actual interface{}) (success bool, err error) {
+	u, err := helper.ToUnstructured(actual)
```
Comment (akrejcir) [inline tests/framework/matcher/restarted.go:26]: Do you expect other types of `actual` than `VirtualMachineInstance`? If not, maybe you can change this to directly typecast to VMI and use it, instead of using unstructered.

File: tests/framework/matcher/restarted.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	}
+	return uid != string(r.oldUID) && phase == string(v1.Running), nil
+}
+
+func (r restartedMatcher) FailureMessage(actual interface{}) (message string) {
+	u, err := helper.ToUnstructured(actual)
+	if err != nil {
+		return err.Error()
+	}
+	return fmt.Sprintf("Expected VMI to be restarted, but got '%v'", u)
```
Comment (akrejcir) [inline tests/framework/matcher/restarted.go:46]: Can you use `format.Message()` as is used in other gomega matchers?

File: tests/framework/matcher/unstructured_bool.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/onsi/gomega/types"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"kubevirt.io/kubevirt/tests/framework/matcher/helper"
+)
+
+func MatchUnstructuredBool(fields ...string) types.GomegaMatcher {
```
Comment (akrejcir) [inline tests/framework/matcher/unstructured_bool.go:14]: What do you think about creating a generic matcher that would use unstructured to access a field, and then run another matcher on that field. Similarly to the `WithTransformMatcher` from gomega.
It could have signature like:
```go
func Field(patch []string, matcher types.GomegaMatcher) types.GomegaMatcher { ... }
```
And be used like:
```go
Expect(vmi).To(Field([]string{"status", "phase"}, Equal("Running")))
```

Comment (0xFelix) [reply tests/framework/matcher/unstructured_bool.go:14]: I like this idea! Will post an updated PR soon.

File: tests/framework/matcher/restarted.go (hunk 8 above)
Comment (akrejcir) [reply tests/framework/matcher/restarted.go:46]: Sure. You can take a look at implementation of any matcher in gomega. For example this is `Equal()`:

https://github.com/kubevirt/kubevirt/blob/25a4b67be739c9634898597ee958ee10ec718529/vendor/github.com/onsi/gomega/matchers/equal_matcher.go#L30-L38

And here is the `format.Message()` definition: 

https://github.com/kubevirt/kubevirt/blob/2f2b42e0d90a181a8fd5bc39aa43b194c895014f/vendor/github.com/onsi/gomega/format/format.go#L68-L87

Comment (akrejcir) [reply tests/framework/matcher/restarted.go:46]: But note that the string representation of `actual` will be trunctuated to a maximum length `fromat.MaxLength`, which is 4000 by default.

**Comment by 0xFelix** [comment]
@akrejcir @rmohr @xpivarc I replaced `MatchUnstructuredBool` and `MatchUnstructuredSlice` with `Field(matcher types.GomegaMatcher, fields ...string)`. WDYT?

**Review by xpivarc — COMMENTED** [review]
Would you be able to split the first commit? I made it halfway and give up.

File: tests/vm_test.go [hunk 10]
```diff
@@ -137,51 +132,29 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
 		})
 	})
 
 	Context("A valid VirtualMachine given", func() {
 		type vmiBuilder func() (*v1.VirtualMachineInstance, *cdiv1.DataVolume)
 
 		newVirtualMachineInstanceWithContainerDisk := func() (*v1.VirtualMachineInstance, *cdiv1.DataVolume) {
-			vmiImage := cd.ContainerDiskFor(cd.ContainerDiskCirros)
-			return tests.NewRandomVMIWithEphemeralDiskAndUserdata(vmiImage, "echo Hi\n"), nil
+			return libvmi.NewCirros(), nil
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tests/vm_test.go:179]: This function is needed in some tables below and IMO it is cleaner to define it once instead of defining the same lambda thrice.

File: tests/vm_test.go [hunk 11]
```diff
@@ -192,68 +165,47 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
 
 		startVM := func(vm *v1.VirtualMachine) *v1.VirtualMachine {
-			By("Starting the VirtualMachine")
+			By("Waiting for VM to exist")
+			Eventually(ThisVM(vm), 300*time.Second, 1*time.Second).Should(Exist())
 
-			Eventually(func() error {
-				updatedVM, err := virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})
+			By("Starting the VirtualMachine")
+			err = tests.RetryWithMetadataIfModified(vm.ObjectMeta, func(meta k8smetav1.ObjectMeta) error {
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tests/vm_test.go:206]: It seemed like the cleaner solution to me. The Eventually blocks with the client usually workaround the same issue where the object was modified.

File: tests/vm_test.go [hunk 12]
```diff
@@ -638,112 +419,78 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
 
-			expectedVMRevisionName := fmt.Sprintf("revision-start-vm-%s-%d", newVM.UID, newVM.Generation)
+			expectedVMRevisionName := fmt.Sprintf("revision-start-vm-%s-%d", vm.UID, vm.Generation)
 			Expect(vmi.Status.VirtualMachineRevisionName).To(Equal(expectedVMRevisionName))
 
-			cr, err := virtClient.AppsV1().ControllerRevisions(newVM.Namespace).Get(context.Background(), vmi.Status.VirtualMachineRevisionName, k8smetav1.GetOptions{})
+			cr, err := virtClient.AppsV1().ControllerRevisions(vm.Namespace).Get(context.Background(), vmi.Status.VirtualMachineRevisionName, k8smetav1.GetOptions{})
 			Expect(err).ToNot(HaveOccurred())
-			Expect(cr.Revision).To(Equal(int64(1)))
+			Expect(cr.Revision).To(Equal(int64(2)))
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tests/vm_test.go:581]: When the startVM func is used the generation is increased by one because the func modifies the VM.

File: tests/vm_test.go [hunk 13]
```diff
@@ -192,68 +165,47 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
-			template := tests.NewRandomVMIWithEphemeralDiskAndUserdata(vmiImage, "echo Hi\n")
-
-			var newVM *v1.VirtualMachine
-			var err error
-
-			newVM = NewRandomVirtualMachineWithRunStrategy(template, runStrategy)
-
-			newVM, err = virtClient.VirtualMachine(util.NamespaceTestDefault).Create(newVM)
+			newVM := tests.NewRandomVirtualMachine(libvmi.NewCirros(
+				libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tests/vm_test.go:192]: I only added it where necessary so the tests would still pass.

File: tests/vm_test.go [hunk 14]
```diff
@@ -192,68 +165,47 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
+			newVM, err := virtClient.VirtualMachine(util.NamespaceTestDefault).Create(newVM)
 			Expect(err).ToNot(HaveOccurred())
-
 			return newVM
 		}
 
 		startVM := func(vm *v1.VirtualMachine) *v1.VirtualMachine {
-			By("Starting the VirtualMachine")
+			By("Waiting for VM to exist")
+			Eventually(ThisVM(vm), 300*time.Second, 1*time.Second).Should(Exist())
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tests/vm_test.go]: Dropped.

**Comment by 0xFelix** [comment]
@xpivarc Updated and rebased the PR and tried to split the changes into more commits. It is still massive, but hopefully the splitted commits make it easier to review?

**Comment by xpivarc** [comment]
@EdDev Do you want to have a look first?

**Review by dankenigsberg — APPROVED** [review]

**Comment by 0xFelix** [comment]
Ping @xpivarc @EdDev 
/retest

**Comment by EdDev** [comment]
> Ping @xpivarc @EdDev /retest

I will try to go over this proposal this week, sorry for the delay in the response.
/cc

**Review by EdDev — COMMENTED** [review]
Thank you for this cleanup! Great work!

A few inline notes added, nothing major.

Here are some general notes:
- I am not sure what `chore` is in the commit topic. I think it can be dropped, leaving just the `tests, vm:` part.
- Some commit subjects (first line) is too long. (line is getting broken when you read it in the review).
- This change is very big with a lot of details, making it hard to review and has a potential for reviewers to miss issues. I think in this case it could have been splitten into smaller PR/s. Please do consider this in the future to help it get in faster.

File: tests/vm_test.go [hunk 15]
```diff
@@ -83,6 +83,22 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 		virtClient = kubevirt.Client()
 	})
 
+	createVM := func(template *v1.VirtualMachineInstance) *v1.VirtualMachine {
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tests/vm_test.go]: Agreed, updated the PR to avoid closures.

File: tests/vm_test.go [hunk 16]
```diff
@@ -1973,13 +1970,19 @@ status:
 			vmRunningRe = regexp.MustCompile("Phase.*Running")
 		})
 
-		It("[test_id:243][posneg:negative]should create VM only once", func() {
-			vm := tests.NewRandomVirtualMachine(libvmi.NewAlpine(), true)
+		generateVMJson := func(running bool) (*v1.VirtualMachine, string) {
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tests/vm_test.go]: ACK, updated the PR to rename it to `createVMAndGenerateJSON`.

**Review by 0xFelix — COMMENTED** [review]
@EdDev Thanks for the review!

I updated the commit subjects to shorten them and remove `chore`. It was inspired by [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

And yes, agreed, next time I will definitely try to keep PRs smaller.

File: tests/vm_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply tests/vm_test.go]: Thanks. The naming is clear now.
I still think that we should not have functions doing to things, if possible, but I won't insist here.

**Review by EdDev — APPROVED** [review]
Thank you!
